SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl

```diff
@@ -1157,10 +1160,12 @@ void MechanicalObject<DataTypes>::init()
     if( x_wA.size() <= 1 && v_wA.size() <= 1 )
     {
         // if a topology is present, implicitly copy position from it
-        if (m_topology != NULL && m_topology->getNbPoints() && m_topology->getContext() == this->getContext())
+        if (l_topology && l_topology->hasPos() /*&& l_topology->getContext() == this->getContext()*/ )
```

Comment:
The test: ./Flexible_test --gtest_filter=TetrahedronVolumeMappingTest/0.test_perTetra
was failing because of this line and line 1090: 
The mapping_test is creating an input mechanicalObject with 4 dofs and an output mechanicalObject with 1 dof only. 
But only one MeshTopology is created in the parent node, with 4 points. thus at the init of the output mechanicalObject was looking for a topology using: l_topology.set( this->getContext()->getActiveMeshTopology() ); 
which returns the parent topology with 4 points instead of 1.

So is it normal that this line doesn't check anymore the context? To fix the test, I set in the output mechanicalObject d_useTopology.setValue(false);
@JeremieA or @fjourdes could you confirm it is the right way of fixing this issue.

Reply:
[ci-build]

Reply:
Could you check that if in `TetrahedronVolumeMapping::apply` the output vector is resized properly ( ie by taking into account the value of `d_volumePerNodes` data, much like what is done during `init`) the crash is removed ?

Reply:
Could you check that if in `TetrahedronVolumeMapping::apply` the output vector is resized properly ( ie by taking into account the value of `d_volumePerNodes` data, much like what is done during `init`) the crash is removed ?

Reply:
asking 3 times, won't make me work faster ;)

I think that in trunk and in this PR the line 1090/1093 return pointer to the MeshTopology of the parent Node (input topology of the mapping).
What change is indeed that the context of the MeshTopology is not checked anymore in line 1163. Uncommenting that test restore the trunk behavior. I.e the mapping_test pass into the else.

I'll check the d_volumePerNodes

Reply:
Intersetingly I could remove the triple comment, but not my wrong edit review text, which was supposed to stay as a comment. 
Did not know either why it was interseted 3 times.
Just as a reminder since now they all got removed.

Could you check that if in `TetrahedronVolumeMapping::apply` the output vector is resized properly ( ie by taking into account the value of `d_volumePerNodes` data, much like what is done during `init`) the crash is removed ?

Reply:
Ok I checked the `TetrahedronVolumeMapping::apply` value of `dOut`
In the test if `d_useTopology.setValue(false);`, size is 1. if not, test failed and size is 4. 
Same if I uncomment `m_topology->getContext() == this->getContext()` size is 1.

So verdict?

Reply:
I do not understand why a mstate should be build from a parent topology?
It is luck that only TetrahedronVolumeMapping_test complains, but almost every situation should crash.
If you want to clean something, remove the implicit link to topology and force to explicitly give the topology to be init with: `<MObject topology="@../../topo"/>`, or even better remove topology from mstate and explicit everything: `<MObject position="@topo.position"/>` but this would break a lot of scenes.

Reply:
I agree with matthieu that we should prefer promote the use of explicit link :)

Reply:
ok but do you want to do that in that PR or do we open another issue? 
If the test is restored, the behavior is similar to the trunk and we can merge the other fix.